src/xhr.js

```diff
@@ -1,208 +1,222 @@
+        pollDelay: 1000,
+        headers: {
+            'Accept': 'application/json; charset=utf-8'
         }
+    }, settings);
+}
+
+/**
+ * Same api as jQuery.ajax - arguments (url, settings) or (settings) with url inside
+ * Additionally content type is automatically json, and object in settings.data is converted to string
+ * to be consumed by GDC backend.
+
+ * settings additionally accepts keys: unathorized, pollDelay  (see xhrSetup for more details)
+ * @method ajax
+ * @param url request url
+ * @param settings settings object
+ */
+export function ajax(url, settings) {
+    let finalSettings;
+    let finalUrl;
+    if ($.isPlainObject(url)) {
+        finalSettings = url;
+        finalUrl = undefined;
+    } else {
+        finalUrl = url;
+        finalSettings = settings;
+    }
+    // copy settings to not modify passed object
+    // settings can be undefined, doesn't matter, $.extend handle it
+    finalSettings = $.extend({}, xhrSettings, finalSettings);
+    if (finalUrl) {
+        finalSettings.url = finalUrl;
+    }
+
+    if ($.isPlainObject(finalSettings.data)) {
+        finalSettings.data = JSON.stringify(finalSettings.data);
+    }
+
+    /*eslint-disable new-cap*/
+    const d = $.Deferred();
+    /*eslint-enable new-cap*/
+    reattachCallbackOnDeferred(finalSettings, 'success', d.done);
+    reattachCallbackOnDeferred(finalSettings, 'error', d.fail);
+    reattachCallbackOnDeferred(finalSettings, 'complete', d.always);
+
+    if (tokenRequest) {
+        continueAfterTokenRequest(finalSettings, d);
+        return d;
+    }
 
-        if ($.isPlainObject(settings.data)) {
-            settings.data = JSON.stringify(settings.data);
+    $.ajax(enrichSettingWithCustomDomain(finalSettings, config.domain)).fail(function jqAjaxFail(xhrObj, textStatus, err) {
+        if (xhrObj.status === 401) {
+            handleUnauthorized(finalSettings, d);
+        } else {
+            d.reject(xhrObj, textStatus, err);
         }
-
-        var d = $.Deferred();
-        reattachCallbackOnDeferred(settings, 'success', d.done);
-        reattachCallbackOnDeferred(settings, 'error', d.fail);
-        reattachCallbackOnDeferred(settings, 'complete', d.always);
-
-        if (tokenRequest) {
-            continueAfterTokenRequest(settings, d);
-            return d;
+    }).done(function jqAjaxDone(data, textStatus, xhrObj) {
+        if (xhrObj.status === 202 && !finalSettings.dontPollOnResult) {
+            // if the response is 202 and Location header is not empty, let's poll on the new Location
+            const location = xhrObj.getResponseHeader('Location');
+            if (location) {
+                finalSettings.url = location;
+            }
+            finalSettings.method = 'GET';
+            delete finalSettings.data;
+            handlePolling(finalSettings, d);
+        } else {
+            d.resolve(data, textStatus, xhrObj);
         }
+    });
+    return d;
+}
+
+// TODO refactor below to use generator
```

Comment:
this is not actionable, I don't understand how this code would benefit from using a generator or what is the value. chances are it will never happen.